sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs

```diff
@@ -76,41 +98,135 @@ private static void CheckForRedundantConditions(ExplodedGraph explodedGraph, Syn
+                var expression = constantExpression as ExpressionSyntax;
+                if (expression != null)
+                {
+                    unreachabExpressions.Add(expression);
+                }
+            }
+
+            var statement = GetUnreachableStatement(unreachabExpressions, constantValue);
+            if (statement != null)
             {
-                context.ReportDiagnostic(Diagnostic.Create(rule, alwaysTrue.GetLocation(), "true"));
+                yield return statement.GetLocation();
             }
+        }
 
-            foreach (var alwaysFalse in conditionFalse.Except(conditionTrue))
+        private static TextSpan GetSpan(SyntaxNode startNode, SyntaxNode endNode)
+        {
+            if (startNode.Equals(endNode))
             {
-                context.ReportDiagnostic(Diagnostic.Create(rule, alwaysFalse.GetLocation(), "false"));
+                return startNode.Span;
             }
+            return TextSpan.FromBounds(startNode.SpanStart, endNode.Span.End);
+        }
+
+        private static SyntaxNode GetUnreachableStatement(IEnumerable<ExpressionSyntax> unreachableExpressions, bool constantValue)
+        {
+            var parent = unreachableExpressions
+                .Select(SyntaxHelper.GetSelfOrTopParenthesizedExpression) // unreachable expressions
+                .Select(node => node.Parent is BinaryExpressionSyntax
+                    ? node.Parent.Parent
+                    : node.Parent) // Constant node is the only expression in an if statement condition
+                .FirstOrDefault();
+
+            if (constantValue)
+            {
+                return (SyntaxNode)
+                    (parent as ConditionalExpressionSyntax)?.WhenFalse ??
+                    (parent as IfStatementSyntax)?.Else?.Statement;
+            }
+            else
+            {
+                return (SyntaxNode)
+                    (parent as ConditionalExpressionSyntax)?.WhenTrue ??
+                    (parent as IfStatementSyntax)?.Statement ??
+                    (parent as WhileStatementSyntax)?.Statement ??
+                    (parent as DoStatementSyntax)?.Statement;
+            }
+        }
+
+        private static IEnumerable<ExpressionSyntax> GetUnreachableExpressions(SyntaxNode constantExpression, bool constantValue)
+        {
+            foreach (var current in constantExpression.AncestorsAndSelf())
+            {
+                if (current.Parent is ParenthesizedExpressionSyntax)
+                {
+                    continue;
+                }
+
+                var binary = current.Parent as BinaryExpressionSyntax;
+                if (!IsShortcuttingExpression(binary, constantValue))
+                {
+                    break;
+                }
+
+                if (binary.Left == current)
+                {
+                    yield return binary.Right;
+                }
+            }
+        }
+
+        private static bool IsLeft(SyntaxNode node)

```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this unused private member. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1144)